openvdb/points/AttributeSet.h

```diff
@@ -201,6 +201,35 @@ class OPENVDB_API AttributeSet
                                         const bool constantStride,
                                         const AttributeArray::ScopedRegistryLock* lock);
 
+    /// @brief Remove and return an attribute array by name
+    /// @param name the name of the attribute array to release
+    /// @details Detaches the attribute array from this attribute set and returns it.
+    /// This also updates the descriptor to remove the reference to the attribute array.
+    /// @note AttributeArrays are stored as shared pointers, so they are not guaranteed
+    /// to be unique. Check the reference count before blindly re-using in a new AttributeSet.
+    AttributeArray::Ptr removeAttribute(const Name& name);
+
+    /// @brief Remove and return an attribute array by index
+    /// @param pos the position index of the attribute to release
+    /// @details Detaches the attribute array from this attribute set and returns it.
+    /// This also updates the descriptor to remove the reference to the attribute array.
+    /// @note AttributeArrays are stored as shared pointers, so they are not guaranteed
+    /// to be unique. Check the reference count before blindly re-using in a new AttributeSet.
+    AttributeArray::Ptr removeAttribute(const size_t pos);
+
+    /// @brief Remove and return an attribute array by index (unsafe method)
+    /// @param pos the position index of the attribute to release
+    /// @details Detaches the attribute array from this attribute set and returns it.
+    /// In cases where the AttributeSet is due to be destroyed, a small performance
+    /// advantage can be gained by leaving the attribute array as a nullptr and not
+    /// updating the descriptor. However, this leaves the AttributeSet in an invalid
+    /// state making it unsafe to call any methods that implicitly derefence the attribute array.
+    /// @note AttributeArrays are stored as shared pointers, so they are not guaranteed
+    /// to be unique. Check the reference count before blindly re-using in a new AttributeSet.
+    /// @warning Only use this method if you're an expert and know the risks of not
+    /// updating the array of attributes or the descriptor.
+    AttributeArray::Ptr removeAttributeUnsafe(const size_t pos);
```

Comment:
Hmm not a huge fan of this 'unsafe' method but I don't see any way around it. I assume you're already using this for some of the point moving changes, is there no way this can be done manually? The only other thing I can suggest is calling `AttributeArray::collapse` (or `AttributeSet::replace`) instead where you're using this which is a single virtual call + the reallocation of a single element - I'd hope that the performance differences would be insignificant vs this

Reply:
Okay I see what you mean. `Grid::getUnsafeAccessor` doesn't leave any objects in an "invalid" state which is the main thing I'm concerned about with this method. Do you have an example of what does the book-keeping looks like with the vector method?

Reply:
> Okay I see what you mean. Grid::getUnsafeAccessor doesn't leave any objects in an "invalid" state which is the main thing I'm concerned about with this method.

I don't agree with this statement. If you delete a node from the tree that the "unsafe" accessor is referencing, the internal cache can still retain a pointer to the deleted node, so any attempt to dereference it will crash. The Accessor _is_ in an invalid state is it not?

> Do you have an example of what does the book-keeping looks like with the vector method?

Sure. If you were to do per-attribute iteration, let's say you wanted to steal all the "N" attribute arrays for a collection of leaf nodes, you'd not only need to extract all the attributes into one vector per-leaf, but you'd also need to keep a reference to each descriptor of each leaf node as it's not guaranteed that the "N" attribute maps to the same index for each leaf. Then you'd look up into this temporary set of vectors/descriptors to find each attribute array.

The alternative with the proposed method is to just extract each attribute array as you needed it, knowing that the leaf node would be destroyed at the end anyway.

Reply:
> I don't agree with this statement. If you delete a node from the tree that the "unsafe" accessor is referencing, the internal cache can still retain a pointer to the deleted node, so any attempt to dereference it will crash. The Accessor is in an invalid state is it not?

There is are differences between the guarantees of the unsafe accessor and this method, but the main point is the accessor does not invalidate the tree. When we talk about what's in an "invalid state", yes I guess you could argue that the accessor could be with your example (though that's no different to say `auto& t = g.tree(); g.newTree(); t.oops();` and actually it's a quirk of the "safe" accessors that this works...). However your method leaves the VDB **data structure** in an invalid state which is, I would argue, a significantly different problem.

> If you were to do per-attribute iteration, let's say you wanted to steal all the "N" attribute arrays for a collection of leaf nodes, you'd not only need to extract all the attributes into one vector per-leaf, but you'd also need to keep a reference to each descriptor of each leaf node as it's not guaranteed that the "N" attribute maps to the same index for each leaf.

Surely in this example you just store N separately or move it to the front/back or w/e, there's no need to keep a reference to the descriptors? I do appreciate it's slightly more complex. Is there any way this method could be made private with friend classes or some such? If not, perhaps this is one for the TSC?

Reply:
In the case where you're stealing attribute arrays in multiple passes (and some of them might not exist), you'd still need to handle all the indexing yourself.

Thanks, I don't have a better solution to propose, so would be good if we can merge this as it is. I'd be happy to revisit when it's clearer how this will be used.